swarm/metrics.py

```diff
@@ -0,0 +1,165 @@
+    is required.
+
+    In a typical case, you might want to find when a loss is below a suitable threshold, eg .05.
+    An alternative might be that you want to find when the range of standard deviations at all x values has
+    falled below a threshold. (This could indicate knots/nodes eliminated.) Or the median/mean/max standard deviation
+    has fallen below a threshold. (This would indicate a level of coherence within the swarm.)
+
+    The threshold can have two different interpretations, either as a nominal value or a ratio relative to a reference
+    value at a given iteration (often the first).  For different regimes of target functions the nominal evaluation score
+    may vary widely widely in terms of magnitude. So the threshold 0.05 can be used to indicate that the score reaches
+    5% of whatever the score was at the first epoch/batch. For ascending scores, a value greater than one may be
+    helpful. i.e. 20 would indicate improving starting value 20 times over.
+
+    This function also allows you to set a criterion, based on whether it should return the first epoch/batch at which
+    the threshold was crossed, or the first time it was crossed and never thereafter exceeded. (I.e. always remains
+    below that threshold in all tested epochs/batches.)
+    Args:
+        epoch_scores: A list or 1-d Numpy array of floats representing the scores at each epoch/catch.
+        threshold: The nominal value you want the score to first get below or above
+        threshold_type: "ratio" or "nominal". Indicate whether you would like to use the threshold as a nominal value, or a ratio that
+            reflects the change from a reference_iteration. Defaults to "nominal".
+        criterion: "first" or "always". Whether you want to return the epoch at which this threshold is first crossed, or
+            always remains below/above (defaults to first)
+        direction: "above" or "below". Choose whether you want get below or above the threshold (defaults below)
+        reference_iteration: Integer for the epoch/batch at which you want to set your reference to be improved upon.
+            Defaults to 0. Ignored if threshold_type is "nominal".
+
+    Returns: An integer specifying which epoch/batch meets this criterion
+
+    """
+    # First check for suitable inputs. I wish there was a neater way of doing this.
+    valid_threshold_types = {"nominal", "ratio"}
+    valid_criteria = {"first", "always"}
+    valid_directions = {"above", "below"}
+    if threshold_type not in valid_threshold_types:
+        raise ValueError("argument `threshold_type` must be one of %r." % valid_threshold_types)
+    if criterion not in valid_criteria:
+        raise ValueError("argument `criterion` must be in one of %r." % valid_criteria)
+    if direction not in valid_directions:
+        raise ValueError("argument `direcction` must be in one of %r." % valid_directions)
+
+    # Convert to numpy for convenience
+    if type(epoch_scores) == list:
```

Comment:
`isinstance(epoch_scores, list)` is the preferred way (as it would handle subclassing).

Also you can just do the `np.array(scores)` since if it fails, it'd just throw an error.

Do you want to handle `torch.Tensor`s?

Reply:
Ah thanks. Good suggestions.
Not sure about `torch.Tensor`s... What do you think? I've barely used them except for in this project. The seem to convert nicely to numpy, so I've never encountered to do anything else with them, since I've been mostly treating them on the way out of an actually neural network training, not on the way in.  (so far). Open to suggestions!

Reply:
I do convert tensors into ndarray's in swarm since we don't need tensors once calculated and they have a lot of metadata for the autograd. I think it's fine to ignore them, only brought it up since you felt the need to handle list too.